test/mem.test.js

```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
+        const app = makeSenecaForTest()
+
+        beforeEach(fin => {
+          app.make('product')
+            .data$({ label: null, price: '3.95' })
+            .save$(fin)
+        })
+
+        beforeEach(fin => {
+          app.make('product')
+            .data$({ label: 'CS101 textbook', price: '134.95' })
+            .save$(fin)
+        })
+
+        it('updates the matching document', fin => {
+          app.test(fin)
+
+          app.ready(() => {
+            app.make('product')
+              .data$({ label: null, price: '5.95' })
+              .save$({ upsert$: ['label'] }, err => {
+                if (err) {
+                  return fin(err)
+                }
+
+                app.make('product').list$({}, (err, products) => {
+                  if (err) {
+                    return fin(err)
+                  }
+
+                  expect(products.length).to.equal(2)
+
+                  expect(products[0]).to.contain({
+                    label: null,
+                    price: '5.95'
+                  })
+
+                  expect(products[1]).to.contain({
+                    label: 'CS101 textbook',
+                    price: '134.95'
+                  })
+
+                  return fin()
+                })
+              })
+          })
+        })
+      })
+
+      describe('when some of the fields in the data$/upsert$ directives do not exist in a document/record', () => {
+        const app = makeSenecaForTest()
+
+        beforeEach(fin => {
+          app.make('product')
+            .data$({ label: 'a toothbrush', price: '3.40' })
+            .save$(fin)
+        })
+
+        it('creates a new document because it can never match any of the documents/records because the none of them have the queried-for fields', fin => {
+          app.test(fin)
+
+          app.make('product')
+            .data$({ label: 'a toothbrush', price: '2.95', coolness_factor: '0.95' })
+            .save$({ upsert$: ['label', 'coolness_factor'] }, err => {
+              if (err) {
+                return fin(err)
+              }
+
+              app.make('product').list$({}, (err, products) => {
+                if (err) {
+                  return fin(err)
+                }
+
+                expect(products.length).to.equal(2)
+
+                expect(products[1]).to.contain({
+                  label: 'a toothbrush',
+                  price: '2.95',
+                  coolness_factor: '0.95'
+                })
+
+
+                return fin()
+              })
+            })
+        })
+      })
+
+      describe('when some of the fields in the upsert$ directive do not exist in the attributes passed via data$', () => {
+        const app = makeSenecaForTest()
+
+        beforeEach(fin => {
+          app.make('product')
+            .data$({ label: 'a toothbrush', price: '3.40' })
+            .save$(fin)
+        })
+
+        it('ignores the missing fields and updates the matching existing record', fin => {
```

Comment:
@rjrodger Do I understand the request correctly - do you mean, if a field is present in the `upsert$` array passed to the `#save$` method, but not inside the object passed to the `#data$` method, - then we should create a new record?

Reply:
It depends what's in the db:
```
db=[{id:1,x:1,y:2,z:3}]  // note: resets for each below

save {x:1,y:22} upsert$:x => db=[{id:1,x:1,y:22,z:3}]   // update as upsert fields match
save {x:11,y:22} upsert$:x => db=[{id:1,x:1,y:22,z:3},{id:2,x:11,y:22}]   // insert as upsert fields do not match

save {x:1,y:2,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33}]   // update as upsert fields match
save {x:11,y:2,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33},{id:2,x:11,y:2,z:33}]   // insert as upsert fields do not match

// undef fields
save {x:1,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33},{id:2,x:1,z:33}] // insert as upsert fields do not match (undef != 2)
```

Update is strict - all upsert fields must match exactly - since undefineds should never match, you can short circuit to insert if any of the upsert fields are undef in the entity to save.

Another way to think about this is that if an ent has no id, insert is the default, and strongly preferred, and update has to  really work hard to happen

Reply:
Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L626